### ISSUE TITLE
[Feature Request][security] Support for Bittorrent v2 format / BitErrant attack

### ISSUE THREAD

**Issue by hannob** [transmission/transmission#458]
I haven't seen any info on whether there's any effort in transmission to support the bittorrent v2 format [1].

Are there any plans or efforts to support it yet? The major reason is the breakage of SHA1, which causes vulnerabilities like the BitErrant attack [2]. The v2 format switches to SHA256 and a Merkle tree structure.

As long as this is unfixed torrents can't be considered as a secure download method that guarantee data integrity.

[1] http://www.bittorrent.org/beps/bep_0052.html
[2] https://biterrant.io/

**Comment by lachesis** [comment]
libtorrent announced BEP-0052 / BitTorrent v2 support today.
https://blog.libtorrent.org/2020/09/bittorrent-v2/

**Comment by da2x** [comment]
Transmission fails to open torrent files encoded with the [v2 reference implementation](https://www.bittorrent.org/beps/bep_0052_torrent_creator.py) that contain v1 and v2 data. The libtorrent blog post suggests this torrent should have been compatible with older clients. It works in a lot of other clients. The error message says “invalid or corrupt torrent file.”

**Comment by sbrudenell** [comment]
SHA-1 chosen-prefix attacks are now in the [USD 10k-100k range](https://eprint.iacr.org/2020/014.pdf). This presents a very real threat, for example against bittorrent-distributed software that might be used by dissidents or whistleblowers.

BitTorrent v2 should be high priority now.

**Comment by ValZapod** [comment]
> real threat, for example against bittorrent-distributed software that

What will happen if Info Hash is generated in DHT that will be the same for different data, i.e. a SHA1 collision? There were some answers on stackechange, but they were not adequate for me...

**Comment by sbrudenell** [comment]
> What will happen if Info Hash is generated in DHT that will be the same for different data, i.e. a SHA1 collision? There were some answers on stackechange, but they were not adequate for me...

[BitErrant](https://www.bleepingcomputer.com/news/security/sha1-collision-attack-can-serve-backdoored-torrents-to-track-down-pirates/) could be used to initially seed a benign payload, then later switch to a malicious payload. You could either selectively deliver the malicious payload only to chosen targets, or just start delivering it to everyone after some time.

**Comment by txtsd** [comment]
Any updates to this? Can we expect it this year?

**Comment by Tykov** [comment]
> 
> 
> Transmission fails to open torrent files encoded with the [v2 reference implementation](https://www.bittorrent.org/beps/bep_0052_torrent_creator.py) that contain v1 and v2 data. The libtorrent blog post suggests this torrent should have been compatible with older clients. It works in a lot of other clients. The error message says “invalid or corrupt torrent file.”

At least it should be fixed

**Comment by spencermp** [comment]
It would be nice to see this feature implimented, if not for massive levels of current usage then for futureproofing.

**Comment by qu1ck** [comment]
Yeah, at least compatibility with hybrid torrents is crucial. Full v2 support can come later but right now transmission just can't open files.

**Comment by ValZapod** [comment]
> merkle tree support

So only qBittorent (does it use libtorrent?) supports applying v2 stuff in network?

**Comment by qu1ck** [comment]
> Yeah, at least compatibility with hybrid torrents is crucial. Full v2 support can come later but right now transmission just can't open files.

I recompiled from current master (70fbce71ec0d) and hybrid files can be opened, thanks.

**Comment by ValZapod** [comment]
> files can be opened

But does v2 protocol work?

**Comment by Skaronator** [comment]
No, only hybrid torrents that contain v1 and v2 information work on the master branch right now. Transmission will parse the v1 data and will ignore v2 data.

**Comment by woojh3690** [comment]
I do not understand why this feature does not have a high priority. In my experience, current half of the torrent is not able to read the transmission.

**Comment by lachesis** [comment]
Yeah I am moving away from Transmission because of this missing feature. It's a shame because qBittorrent's CLI and web UI are much worse, and the client runs slower than Transmission on my underpowered seedbox, but there have been several torrents in the last few months that I was simply unable to download with Transmission.

**Comment by rebeltaz** [comment]
So this has been an issue for FIVE YEARS!? with no resolution!?

**Comment by joshenders** [comment]
> So this has been an issue for FIVE YEARS!? with no resolution!?

If it bothers you so much, why not try submitting a PR?

**Comment by rebeltaz** [comment]
Because I don't know what a PR is?

**Comment by Zipdox** [comment]
> No, only hybrid torrents that contain v1 and v2 information work on the master branch right now. Transmission will parse the v1 data and will ignore v2 data.

I created a hybrid torrent in qBitTorrent and it's still counted as corrupt. Adding using the magnet link just gets stuck in a loop of downloading metadata.

**Comment by ckerr** [comment]
> I created a hybrid torrent in qBitTorrent and it's still counted as corrupt. Adding using the magnet link just gets stuck in a loop of downloading metadata.

@Zipdox testing with what Transmission build?

If you're still seeing this in a recent build, can you create a hybrid torrent with public domain contents and attach it to this ticket? No need to seed it; I just want to figure out what's not parsing.

**Comment by Zipdox** [comment]
transmission-daemon 3.00 (bb6b5a062e)

**Comment by Skaronator** [comment]
https://github.com/transmission/transmission/commit/bb6b5a062ee594dfd4b7a12a6b6e860c43849bfd doesn't include https://github.com/transmission/transmission/commit/fb39c4663c9e6b7fadc7cd55bdcfa04167ae53f1 @Zipdox

**Comment by Zipdox** [comment]
Classic Debian stable

**Comment by ValZapod** [comment]
Can be we add here a test file that is not hybrid, i.e. only has v2 data? That will not work, as only hybrid are supported. Can you convert hybrid to v1+v2 separte, like ttc files? Because no Merkle tree support?? Also can somebody paste a sha256 magnet? Will sha1 bt hash magnet hybrid file or only v1 file? Wow. Crazy!

**Comment by kovalensky** [comment]
Is there any bypass for magnet links to work with Transmission?

We are trying to enable BitTorrent v2 support in our torrent tracker, but since Transmission clients having problems we got our hands tied.

**Comment by ckerr** [comment]
@kovalensky depends on what you're asking for:

BitTorrent v2 has the concept of hybrid torrents, where the torrent file & magnet links can have both the sha1 and sha256 info hash. Transmission 4, which is in beta now, supports v2 hybrid torrents and their magnet links.

But it still needs that sha1 hash, so a torrent that is **only** v2 won't work in Transmission until 5.

**Comment by kovalensky** [comment]
> @kovalensky depends on what you're asking for:
> 
> BitTorrent v2 has the concept of hybrid torrents, where the torrent file & magnet links can have both the sha1 and sha256 info hash. Transmission 4, which is in beta now, supports v2 hybrid torrents and their magnet links.
> 
> But Transmission 4 still needs that sha1 hash, so a torrent that is **only** v2 (i.e. not a hybrid) won't work in Transmission until 5.

We enabled Bittorrent v2 support (15k users) base, but 30% of our good seeding users using Transmission version lesser than 4. 
So they are having problems downloading hybrid torrents and magnet links.
We disabled uploading of v2 only torrents, only hybrids and v1 allowed, since that's a great feature to show files' sha2 root hash in torrent's about page.
We don't encourage users to switch to Transmission 4, since it's still in beta.
On the topic of those problems we had to disable hybrid magnet links.
So I was asking if there's a some bypass for this, maybe some additional string in magnet link, idk

**Comment by ckerr** [comment]
Sounds like the solution is to just get 4.0.0 stable out the door

**Comment by ckerr** [comment]
Looks related to #3386. Very frustrating.

**Comment by s777s** [comment]
In 2023 there is no need for hybrid (using v1 and ignoring v2). V2 only will be enough. Stable 3.0 is in repos of many distros now. Will it be implemented in upcoming 4.0? Why you mark it like "feature"? Client can't open torrent files, it is must urgent Bug! :(
Do you understand, that many users of your software (like me), need to look for another client due to BUG, that has a few years history and this BUG makes client nonworking?

**Comment by txtsd** [comment]
@s777s This kind of tone and approach always wears away at the developers. Please try to be nicer next time despite your frustration :+1:

**Comment by Nadahar** [comment]
@s777s While I too would appreciate v2 support, it's very frivolous to call it a bug. It is not.

To make any software support something, somebody must implement it. I'm sure there's quite a lot of work involved in implementing v2 torrents. So, it all comes down to who has the knowledge, will and opportunity to implement it. I very much doubt labeling something a "bug" and going on a little tantrum will help motivate anybody to find the time and invest the effort needed.

**Comment by ckerr** [comment]
There are a couple of different related / overlapping questions here that I want to answer:

- The nightly builds used to support reading v2 torrent files but #3386, which fixed an issue in hybrid files, regressed on that. Fixing that regression is something I'm going to investigate for `4.0.x` but won't decide until I understand how big / risky / intrusive the change would be. 

- Even if this does get fixed in `4.0.x`, things there would work by reading v2 torrent files and using just enough to treat them as v1 torrents. It does not have support for merkle trees, for the new v2 peer protocol messages, or for the v2-style padding. All of that requires new code and by definition cannot go into `4.0.x` because Transmission now follows semver (see #1037) and `4.0.x` should only fix bugs.

- Transmission won't go multiple years between releases anymore. For example `4.0.0` has gotten _far_ more feedback than the betas did so I expect there will be a `4.0.1` within a week or two. Even if this stays as-is in `4.0.x` that doesn't mean we'll be stuck with the status quo forever.

- Shouting at me and threatening to take your business elsewhere is not interesting or motivating. If you must be a Karen, please do take your business elsewhere because it will make you happier and Transmission's community better.

For everyone who wants these changes to come sooner -- requests are welcomed. That's serious, not snark. The maintainers, including me, have made it a priority to do PR review and help contributors the PRs past the finish line.

**Comment by s777s** [comment]
@ckerr to be honest, I appreciate your work and using your client for many years, so I want to say sorry for my tone and thank you for your work. Was a little irritated, because on some trackers there are some seeds (not all) with incompatible newer version. And was a little surprised to see words "vulnerability", "2017" and "feature" in the top of the thread while was googling why torrent file doesn't work.

**Comment by zebediah49** [comment]
So, either I'm very confused here, or I guess should file this as a documentation bug.  The first entry in the 4.0.0 release's New Features changelog section is

>Support for using BitTorrent v2 torrents and hybrid torrents. (Support for creating v2 and hybrid torrents is slated for an upcoming release.)

But this issue seems to be suggesting that it just supports "not breaking", and can't actually use v2 torrents?  That seems highly misleading, to the point where I'm not actually sure what the software can do.

**Comment by manofsticks** [comment]
Reading through this and had some questions. But first the situation to make my questions clearer:

A private tracker I'm on announced they are temporarily restricting Hybrid and V2 torrents, only allowing V1.

Upon creation of a new torrent in Transmission 4.0.1 and uploading it, I received an error on the site. 

Downgrading to Transmission 3.0 and creating the torrent worked fine and upload was successful.

I was under the impression from the notes that Transmission 4.0.1 was still creating V1 torrent files, but simply had support for utilizing them? I assume either A) I am creating the torrent incorrectly based on some new setting in 4.0.1, B) The private tracker made a mistake in their Hybrid/V2 bans, or C) I am misunderstanding the situation entirely, as I was not aware there were V1 and V2 torrents until earlier today.

Any clarity appreciated

**Comment by hhartzer** [comment]
Should this still be open now that v4.0.0 (and later) is out?

It seems like the only piece missing is the ability to create v2 torrents.

**Comment by revher** [comment]
Yes, no support yet with the example of [https://blog.libtorrent.org/2020/09/bittorrent-v2/](url) 
`$ transmission-remote -a bittorrent-v2-test.torrent`
`Error: unrecognized info`
with 
`$ transmission-cli -v`
`transmission-cli 4.0.5 (a6fe2a64aa)`

**Comment by revher** [comment]
By the way, I compiled Transmission from Git (4.1.0-dev (1855cdb731)) successfully on my Monterey 12.6. Bravo! Support for V2 is not here yet. May be it could be added on the Readme.md that there is no support yet because you can google people having run V2 successfully with version 4.0!

**Comment by PhoenixBlade24** [comment]
I am using Transmission 4.0.6 on Arch Linux, and it is still unable to open V2 torrents. Neither V2 magnet links or standalone torrent files.

**Comment by aryasenna** [comment]
For those like me coming from Google wondering what are the current states (4.0.6 stable also 4.1.x dev), I hope I can save you the trouble of the long read of this 7 (!) years old issue. 

1. Transmission 4.0.x does not (and won't?) support Bittorent v2.
2. It *does* support *hybrid* v1/v2 torrent, but the torrent will be treated as v1 torrent.
3. Related to # 1, this is referring to the maintainer statement that (a) the codes to support v2 are just not there (b) major changes  are to be included in next minor/major version release (5.x.x/4.1.x) and not patch (4.0.x).
4. This is just my understanding, but the maintainer states that the project are not actively working on Bittorrent V2 support (now the 7 years delay makes sense), but will rely on community's PR. In this regard they are 'committed' to help/support the community to make the PR mergeable.

The last point is not ideal but understandable, I  personally appreciate the transparency, and this givessome of us the right trigger to migrate to other client.

**Comment by sojuo** [comment]
> For those like me coming from Google wondering what are the current states (4.0.6 stable also 4.1.x dev), I hope I can save you the trouble of the long read of this 7 (!) years old issue.
> 
>     1. Transmission 4.0.x does not (and won't?) support Bittorent v2.
> 
>     2. It _does_ support _hybrid_ v1/v2 torrent, but the torrent will be treated as v1 torrent.
> 
>     3. Related to # 1, this is referring to the maintainer statement that (a) the codes to support v2 are just not there (b) major changes  are to be included in next minor/major version release (5.x.x/4.1.x) and not patch (4.0.x).
> 
>     4. This is just my understanding, but the maintainer states that the project are not actively working on Bittorrent V2 support (now the 7 years delay makes sense), but will rely on community's PR. In this regard they are 'committed' to help/support the community to make the PR mergeable.
> 
> 
> The last point is not ideal but understandable, I personally appreciate the transparency, and this gives some of us the right trigger to migrate to other client.

I would like to clarify my understanding—are you saying that Transmission [4.0.0](https://github.com/transmission/transmission/tree/4.0.0) and [4.0.1](https://github.com/transmission/transmission/tree/4.0.1) can open torrents in **v2 mode** (by reading their contents) but do not actually support downloading/seeding them using the v2 protocol?

Additionally, will **none of the 4.x.x releases** support v2 torrents? Does Transmission officially have no long-term plans to implement v2 support?

If that is the case, it would be rather disappointing, as v2 torrents reportedly offer efficiency improvements, including reduced power consumption—a significant advantage for many users.

*(Note: I am a non-native English speaker and have tried to phrase this politely and concisely. If any wording could be misconstrued, please disregard potential unintended tone—I mean no offence.)*

**Comment by aryasenna** [comment]
>The [4.0.0](https://github.com/transmission/transmission/tree/4.0.0) and [4.0.1](https://github.com/transmission/transmission/tree/4.0.1) versions can only open v2 torrents to inspect their contents, rather than fully supporting v2 protocol for data transfer.

So .torrent file has the these variants:

- v1-only  torrent based on https://www.bittorrent.org/beps/bep_0003.html
- v2-only .torrent based on https://www.bittorrent.org/beps/bep_0052.html
- v2 hybrid torrent, also based on https://www.bittorrent.org/beps/bep_0052.html, it contains both v1 and v2 fields.

Transmission can currently only open v1-only and v2 hybrid torrents. 
When opening a hybrid torrent file, Transmission will only reads v1 data and does nothing with v2 fields.
Transmission then will download the torrent using v1 protocol.

See an example on: https://github.com/Lanzaa/transmission/blob/88a3149ce4b727515e908ecd58c0027aa0ffd37e/libtransmission/torrent-metainfo.cc#L345

>Additionally, is it accurate that no 4.x.x versions will ever support v2 torrents, and that Transmission’s developers have no long-term plans to implement this feature?

You have to wait for any of maintainers' to answer that kind of question.

**Comment by s777s** [comment]
Maybe you implement v2? Nobody wants to work for free. What done, is done.
Thanks to founder. What needs to be done - nobody have time ot wish to do.

ср, 30 апр. 2025 г., 17:39 Arya Senna ***@***.***>:

**Comment by s777s** [comment]
I'm apologize to my bad words at the beginning. But it is still bug. Don't
know C, only php. Will try to study C, to clear that bug. That can't open
v2 files. Will take many time.

вс, 11 мая 2025 г., 02:53 Stas Stas ***@***.***>:
